cmd/sonobuoy/app/runner.go

```diff
@@ -0,0 +1,330 @@
+			errlog.LogError(errors.New("Preflight checks failed"))
+			for _, err := range errs {
+				errlog.LogError(err)
+			}
+			return true
+		}
+	}
+
+	if err := sonobuoyClient.Run(runCfg); err != nil {
+		errlog.LogError(errors.Wrap(err, "error attempting to run sonobuoy"))
+		return true
+	}
+
+	return false
+}
+
+func retrieve(outDir string) bool {
+	// Get a reader that contains the tar output of the results directory.
+	reader, ec := sonobuoyClient.RetrieveResults(&client.RetrieveConfig{Namespace: runnerflags.namespace})
+
+	eg := &errgroup.Group{}
+	eg.Go(func() error { return <-ec })
+	eg.Go(func() error { return untar(reader, grabTarballResults) })
+
+	err := eg.Wait()
+	if _, ok := err.(exec.CodeExitError); ok {
+		fmt.Fprintln(os.Stderr, "Results not ready yet. Check `sonobuoy status` for status.")
+		return true
+
+	} else if err != nil {
+		fmt.Fprintf(os.Stderr, "error retrieving results: %v\n", err)
+		return true
+	}
+
+	return false
+}
+
+func status() (*aggregation.Status, error) {
+	status, err := sonobuoyClient.GetStatus(statusFlags.namespace)
+	if err != nil {
+		return status, errors.Wrap(err, "error attempting to run sonobuoy")
+	}
+
+	if runnerflags.showAll {
+		err = printAll(os.Stdout, status)
+	} else {
+		err = printSummary(os.Stdout, status)
+	}
+
+	return status, err
+}
+
+func delete() {
+	client, err := sonobuoyClient.Client()
+	if err != nil {
+		errlog.LogError(err)
+		os.Exit(1)
+	}
+
+	rbacEnabled, err := runnerflags.rbacMode.Enabled(client)
+	if err != nil {
+		errlog.LogError(errors.Wrap(err, "couldn't detect RBAC status"))
+		os.Exit(1)
+	}
+	deleteopts.EnableRBAC = rbacEnabled
+
+	if err := sonobuoyClient.Delete(&deleteopts); err != nil {
+		errlog.LogError(errors.Wrap(err, "failed to delete sonobuoy resources"))
+		os.Exit(1)
+	}
+}
+
+func untar(reader io.Reader, fileHandler func(*tar.Header, io.Reader) error) error {
+	tarReader := tar.NewReader(reader)
+	for {
+		header, err := tarReader.Next()
+		if err == io.EOF {
+			break
+		}
+
+		if err != nil {
+			return err
+		}
+
+		if header == nil {
+			continue
+		}
+
+		switch header.Typeflag {
+		case tar.TypeReg:
+			if err := fileHandler(header, tarReader); err != nil {
+				return err
+			}
+		default:
+			continue
+		}
+	}
+
+	return nil
+}
+
+func grabTarballResults(header *tar.Header, tarReader io.Reader) error {
+	writeBuffer := new(bytes.Buffer)
+	writer := bufio.NewWriter(writeBuffer)
+	if _, err := io.Copy(writer, tarReader); err != nil {
+		return err
+	}
+
+	if err := writer.Flush(); err != nil {
+		return err
+	}
+
+	readerBuffer := bytes.NewBuffer(writeBuffer.Bytes())
+	gzipReader, err := gzip.NewReader(readerBuffer)
+	if err != nil {
+		fmt.Println(err)
+		return err
+	}
+	defer gzipReader.Close()
+
+	err = untar(gzipReader, printTarballResults)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func printTarballResults(header *tar.Header, tarReader io.Reader) error {
```

Comment:
are you just printing the untarrred result to stdout?

Reply:
Yes, with each file on a separate line so that is recognized as a single event by logging daemons/drivers. Our current idea is to just use logging as our way to persist results of successful runs.

Reply:
That sounds a bit terrible :confused: 
Kubernetes's own CI would not want to use this, we'd upload the tarball contents separately to persistent storage (GCS) rather than dump everything into the logs.

Reply:
We do not have GCS, it would feel more apt to allow multiple ways to output the result to various backends and not just GCS. At current though we don't really have usage scenarios to even test adding those different persistence stores as backends. At the same time, someone within our team also brought up that it would be nice if we had a way to parse the results to know if there were failures within the runs without having to manually scan the json files.

Reply:
I don't think sonobuoy should upload to GCS, I think your automation should upload the output files to $storage ideally.

Reply:
As for checking if there are failures, the exit code of the command should be sufficient without dumping the results to the logs.